File: core/src/mindustry/client/utils/ProcessorFinder.java
package mindustry.client.utils;

import arc.Core;
import arc.struct.Seq;
import kotlin.text.Regex;
import mindustry.Vars;
import mindustry.client.navigation.clientThread;
import mindustry.gen.Building;
import mindustry.graphics.Drawf;
import mindustry.world.Tile;
import mindustry.world.Tiles;
import mindustry.world.blocks.logic.LogicBlock.LogicBuild;

import java.util.concurrent.CopyOnWriteArrayList;

import static mindustry.Vars.player;
import static mindustry.Vars.tilesize;

public class ProcessorFinder {
    private static final CopyOnWriteArrayList<LogicBuild> highlighted = new CopyOnWriteArrayList<>();
    public static final CopyOnWriteArrayList<Regex> queries = new CopyOnWriteArrayList<>();

    public static void search() {
        highlighted.clear();
        Seq<Building> builds = player.team().data().buildings.filter(b -> b instanceof LogicBuild);
    
        clientThread.post(() -> {
            int matchCount = 0, processorCount = 0;
            for (Building build : builds) {
                var logicBuild = (LogicBuild) build;
                for (Regex query : queries) {
                    if (query.containsMatchIn((logicBuild.code))) {
                        matchCount++;
                        highlighted.add(logicBuild);
                    }
                    processorCount++;
                }
            }
    
            int finalMatchCount = matchCount;
            int finalProcessorCount = processorCount;
            Core.app.post(() -> {
                if (finalMatchCount == 0) player.sendMessage("[accent]No matches found.");
                else player.sendMessage(String.format("[accent]Found [coral]%d/%d[] [accent]matches.", finalMatchCount, finalProcessorCount));
            });
        });
    
    }
    
    public static void searchAll() {
        highlighted.clear();
    
        Tiles tiles = Vars.world.tiles;
        clientThread.post(() -> {
            int matchCount = 0, processorCount = 0;
            for (Tile tile : tiles) {
                if (tile.build instanceof LogicBuild logicBuild) {
                    for (Regex query : queries) {
                        if (query.containsMatchIn((logicBuild.code))) {
                            matchCount++;
                            highlighted.add(logicBuild);
                        }
                        processorCount++;
                    }
                }
            }
    
            int finalMatchCount = matchCount;
            int finalProcessorCount = processorCount;
            Core.app.post(() -> {
                if (finalMatchCount == 0) player.sendMessage("[accent]No matches found.");
                else player.sendMessage(String.format("[accent]Found [coral]%d/%d[] [accent]matches.", finalMatchCount, finalProcessorCount));
            });
        });
    }

    public static void list() {
        StringBuilder sb = new StringBuilder(Core.bundle.get("client.command.procfind.list"));
        for (LogicBuild build : highlighted) sb.append(String.format("(%d, %d)", (int) build.x >> 3, (int) build.y >> 3)).append(", ");
        player.sendMessage(sb.toString());
    }

    public static void clear() {
        queries.clear();
        highlighted.clear();
    }

    public static int getCount() {
        return highlighted.size();
    }

    public static void draw() {
        for (LogicBuild build : highlighted) {
            Drawf.square(build.x, build.y, build.block.size * tilesize / 2f + 8f);
        }
    }
}

File: core/src/mindustry/client/utils/ProcessorFinder.kt
package mindustry.client.utils

import arc.*
import arc.struct.*
import mindustry.Vars.*
import mindustry.client.navigation.*
import mindustry.gen.*
import mindustry.graphics.*
import mindustry.world.blocks.logic.*
import java.util.concurrent.*

object ProcessorFinder {
    private val highlighted: CopyOnWriteArrayList<LogicBlock.LogicBuild> = CopyOnWriteArrayList()
    val queries: CopyOnWriteArrayList<Regex> = CopyOnWriteArrayList()

    fun search(query: Regex) {
        highlighted.clear()
        val builds: Seq<Building> = player.team().data().buildings.filter { it is LogicBlock.LogicBuild }
    
        clientThread.post {
            var matchCount = 0
            var processorCount = 0
            for (build in builds) {
                val logicBuild = build as LogicBlock.LogicBuild
                if (query.containsMatchIn((logicBuild.code))) {
                    matchCount++
                    highlighted.add(logicBuild)
                }
                processorCount++
            }
    
            Core.app.post {
                if (matchCount == 0) player.sendMessage(Core.bundle.get("client.processorpatcher.nomatches"))
                else player.sendMessage(Core.bundle.format("client.processorpatcher.foundmatches", matchCount, processorCount))
            }
        }

        queries.add(query)
    }
    
    fun searchAll() {
        highlighted.clear()
    
        val tiles = world.tiles
        clientThread.post {
            var matchCount = 0
            var processorCount = 0
            for (tile in tiles) {
                if (tile.build is LogicBlock.LogicBuild) {
                    for (query in queries) {
                        if (query.containsMatchIn(((tile.build as LogicBlock.LogicBuild).code))) {
                            matchCount++
                            highlighted.add((tile.build as LogicBlock.LogicBuild))
                        }
                    }
                    processorCount++
                }
            }
    
            Core.app.post {
                if (matchCount == 0) player.sendMessage(Core.bundle.get("client.procesorpatcher.nomatches"))
                else ui.chatfrag.addMsg(Core.bundle.format("client.procesorpatcher.foundmatches", matchCount, processorCount))
            }
        }
    }

    fun list() {
        val sb = StringBuilder(Core.bundle.get("client.command.procfind.list"))
        for (build in highlighted) {
            sb.append(String.format("(%d, %d), ", (build.x / 8).toInt(), (build.y / 8).toInt()))
        }
        ui.chatfrag.addMsg(sb.toString()).findCoords()
    }

    fun clear() {
        queries.clear()
        highlighted.clear()
    }

    fun getCount(): Int {
        return highlighted.size
    }

    fun draw() {
        for (build in highlighted) {
            Drawf.square(build.x, build.y, build.block.size * tilesize / 2f + 8f)
        }
    }
}
